frame/assistants/shadow-liketh/src/shadow.rs

```diff
@@ -0,0 +1,200 @@
+        http: reqwest::Client,
+        bridge: BridgeName,
+    ) -> Self {
+        Self {
+            config,
+            gql,
+            eth,
+            http,
+            bridge,
+        }
+    }
+}
+
+impl Shadow {
+    fn read_graphql(&self, file: impl AsRef<str>) -> ShadowComponentReuslt<&str> {
+        let file = file.as_ref();
+        let dir = self.bridge.directory();
+        let graph = GRAPHQL_DIR
+            .get_file(format!("{}/{}", dir, file))
+            .or_else(|| GRAPHQL_DIR.get_file(format!("generic/{}", file)))
+            .ok_or_else(|| ShadowComponentError::GraphQL("No graphql fround".to_string()))?;
+        graph
+            .contents_utf8()
+            .ok_or_else(|| ShadowComponentError::GraphQL("Failed to read graphql file".to_string()))
+    }
+}
+
+impl Shadow {
+    pub async fn parcel(&self, block_number: u64) -> ShadowComponentReuslt<HeaderParcel> {
+        tracing::debug!(target: "shadow", "Shadow parcel fo block: {}", block_number);
+        let header = self
+            .eth
+            .get_header_by_number(block_number)
+            .await
+            .map_err(|e| ShadowComponentError::Ethereum(format!("{:?}", e)))?;
+        let mmr_root = self.mmr_root(block_number - 1).await?;
+        Ok(HeaderParcel { mmr_root, header })
+    }
+
+    pub async fn receipt(
+        &self,
+        tx: impl AsRef<str>,
+        last: u64,
+    ) -> ShadowComponentReuslt<EthereumReceiptWithMMRProof> {
+        let resp = self
+            .http
+            .get(&format!(
+                "{}/ethereum/receipt/{}",
+                &self.config.endpoint,
+                tx.as_ref(),
+            ))
+            .send()
+            .await?;
+        if resp.status() == reqwest::StatusCode::INTERNAL_SERVER_ERROR {
+            return Err(ShadowComponentError::InternalServer(resp.text().await?));
+        }
+        let result: serde_json::Value = resp.json().await?;
+        if let Some(err) = result.get("error") {
+            tracing::error!(
+                target: "shadow",
+                "Error response by shadow: {:?}",
+                result,
+            );
+            let msg = err.as_str().ok_or_else(|| {
+                ShadowComponentError::Cusom("Failed parse error message".to_string())
+            })?;
+            return Err(ShadowComponentError::Cusom(msg.to_owned()));
+        }
+        let receipt: EthereumReceiptJson = serde_json::from_value(result)?;
+        let header = &receipt.header;
+
+        let (member_leaf_index, last_leaf_index) = (header.number, last);
+        let proof = self.mmr_proof(member_leaf_index, last_leaf_index).await?;
+        let mmr_proof = MMRProofJson {
+            member_leaf_index,
+            last_leaf_index: last_leaf_index - 1,
+            proof,
+        };
+        Ok(EthereumReceiptWithMMRProof { receipt, mmr_proof })
+    }
+}
+
+impl Shadow {
+    fn extract_peaks(&self, positions: Vec<MMRNode>) -> Vec<(u64, [u8; 32])> {
```

Comment:
```suggestion
    fn extract_positions_and_hashs(&self, positions: Vec<MMRNode>) -> Vec<(u64, [u8; 32])> {
```

Reply:
Please wait,. let me update reference after this chagned